### PR TITLE
 add strategy filter

### PR DESCRIPTION
**Pull request by alavieille**
https://github.com/open-orchestra/open-orchestra-libs/pull/11
https://github.com/open-orchestra/open-orchestra-user-bundle/pull/71
https://github.com/open-orchestra/open-orchestra-workflow-function-bundle/pull/39
https://github.com/open-orchestra/open-orchestra-cms-bundle/pull/1019
https://github.com/open-orchestra/open-orchestra-base-api-mongo-model-bundle/pull/6
https://github.com/open-orchestra/open-orchestra-model-bundle/pull/406
https://github.com/open-orchestra/open-orchestra-base-api-bundle/pull/49

### CODE REVIEW

File: Repository/AbstractAggregateRepository.php [hunk 1]
```diff
@@ -12,6 +13,11 @@
 abstract class AbstractAggregateRepository extends DocumentRepository
 {
     /**
+     * @var FilterTypeManager
+     */
+    protected $filterTypeManager;
```
Comment (nicolasThal) [inline Repository/AbstractAggregateRepository.php]: I would have rather put the filterTypeManager declaration in the trait

**Comment by alavieille** [comment]
@nicolasThal

File: Pagination/MongoTrait/FilterTypeStrategy/FilterTypeManager.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    public function generateFilter($type, $name, $value, $documentName)
+    {
+        $filter = null;
+
+        /**
+         * @var FilterTypeInterface $filterType
+         */
+        foreach ($this->filters as $filterType){
+            if ($filterType->support($type)) {
+                $filter = $filterType->generateFilter($name, $value, $documentName);
```
Comment (nicolasThal) [inline Pagination/MongoTrait/FilterTypeStrategy/FilterTypeManager.php]: shouldn't you merge all the filter?

If no, you could directly return the result of the generateFilter method

File: Pagination/Tests/MongoTrait/FilterTypeStrategy/FilterTypeManagerTest.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    public function testGenerateFilter()
+    {
+        $filter = $this->manager->generateFilter('fakeType','fakeName','fakeValue', 'fakeDocumentName');
+        $this->assertSame(null, $filter);
+
+        $this->manager->addStrategy($this->filterStrategie);
+        $filter = $this->manager->generateFilter('fakeType','fakeName','fakeValue', 'fakeDocumentName');
+        $this->assertSame('fakeFilter', $filter);
+    }
+}
```
Comment (nicolasThal) [inline Pagination/Tests/MongoTrait/FilterTypeStrategy/FilterTypeManagerTest.php]: eof

File: Pagination/Tests/MongoTrait/FilterTypeStrategy/Strategies/StringFilterStrategyTest.php [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+ * Class StringFilterStrategyTest
+ */
+class StringFilterStrategyTest extends \PHPUnit_Framework_TestCase
+{
+    /**
+     * @var StringFilterStrategy
+     */
+    protected $strategy;
+
+    public function setUp()
```
Comment (nicolasThal) [inline Pagination/Tests/MongoTrait/FilterTypeStrategy/Strategies/StringFilterStrategyTest.php:20]: method comment

**Comment by alavieille** [comment]
@nicolasThal

File: Pagination/MongoTrait/FilterTypeStrategy/FilterTypeManager.php [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        /**
+         * @var FilterTypeInterface $filterType
+         */
+        foreach ($this->filters as $filterType){
+            if ($filterType->support($type)) {
+                return $filterType->generateFilter($name, $value, $documentName);
+            }
+        }
+
+        return $filter;
```
Comment (nicolasThal) [inline Pagination/MongoTrait/FilterTypeStrategy/FilterTypeManager.php]: Maaaaaaaaaaaaaaaaaaaaaaaaaaaaan, you could have just put `return null;` at the end and not declare the `$filter` property!!!!

**Comment by alavieille** [comment]
:dizzy_face:   Pr updated @nicolasThal

**Comment by nicolasThal** [comment]
:+1:

**Pull request merged**